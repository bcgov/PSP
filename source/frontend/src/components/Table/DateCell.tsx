import { CellProps } from 'react-table';
import { prettyFormatDate, prettyFormatDateTime } from 'utils';

export const DateCell = ({ cell: { value } }: CellProps<Date, string>) => prettyFormatDate(value);

export const DateTimeCell = ({ cell: { value } }: CellProps<Date, string>) =>
  prettyFormatDateTime(value);
