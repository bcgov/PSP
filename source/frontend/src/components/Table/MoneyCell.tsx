import { CellProps } from 'react-table';

import { formatMoney } from '@/utils';

export const MoneyCell = ({ cell: { value } }: CellProps<any, number | '' | undefined>) => (
  <div>{value === undefined || value === '' ? '' : formatMoney(value)}</div>
);
