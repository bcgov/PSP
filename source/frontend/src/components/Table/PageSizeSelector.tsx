import { IMenuItemProps, Menu } from 'components/menu/Menu';
import React from 'react';
import Form from 'react-bootstrap/Form';
import styled from 'styled-components';

/** align text to middle of size input */
const StyledText = styled.span`
  margin-top: 0.3rem;
`;

interface IProps {
  options: number[];
  value: number;
  onChange: (size: number) => void;
  alignTop: boolean;
}

export const TablePageSizeSelector: React.FC<React.PropsWithChildren<IProps>> = ({
  options,
  value,
  onChange,
  alignTop,
}) => {
  const [selected, setSelected] = React.useState(value);

  const handleValueChange = (newSelection: number) => {
    if (newSelection !== selected) {
      setSelected(newSelection);
      onChange(newSelection);
    }
  };

  const pageSizeOptions: IMenuItemProps[] = options.map(option => ({
    label: option,
    value: option,
    onClick: () => handleValueChange(option),
  }));
  return (
    <Menu options={pageSizeOptions} alignTop={alignTop}>
      <div style={{ display: 'flex' }}>
        <StyledText>Show</StyledText>
        {/*the selector appears to capture the click event which prevents the page change from being registered*/}
        <Form.Control
          size="sm"
          value={`${selected}`}
          type="number"
          style={{ width: 50, marginLeft: 10, marginRight: 10, backgroundColor: 'white' }}
          disabled
        />
        <StyledText>Entries</StyledText>
      </div>
    </Menu>
  );
};
