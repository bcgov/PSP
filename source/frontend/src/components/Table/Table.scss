@import '~assets/scss/styles.scss';
@import '~assets/scss/colors.scss';
.table {
  .thead-light {
    .th {
      color: $table-header-text-color;
      background-color: $table-header-bg-color;
    }
  }
  .thead {
    .th {
      border-bottom: 0.2rem solid $table-header-border-color;
      cursor: pointer;
      font-size: 1.6rem;
      font-weight: bold;
      line-height: 2.4rem;
    }
  }
  .tbody {
    .tr-wrapper {
      &:nth-child(even) {
        background-color: $table-even-row-bg-color;
      }
      &:nth-child(odd) {
        background-color: $table-odd-row-bg-color;
        &:last-child {
          border-bottom: 0.1rem solid $table-column-border-color;
        }
      }
      .tr {
        &:hover {
          background-color: $table-hover-color;
        }
        &.selected {
          background-color: $selected-color;
        }
      }
      .tbody {
        padding-left: 5rem;
        .tr-wrapper {
          border-bottom: none;
        }
      }
    }
  }
  .table-loading,
  .no-rows-message {
    display: flex;
    flex-direction: column;
    width: 100%;
    justify-content: center;
    align-items: center;
    margin-top: 0.5rem;
    margin-bottom: 0.5rem;
  }

  .svg-btn {
    color: $text-color;
  }

  .svg-btn:hover {
    color: $icon-light-color;
    text-decoration: none;
    cursor: pointer;
  }

  .svg-btn:hover svg {
    fill: $icon-light-color;
    stroke: #fff;
  }

  .svg-btn.active {
    color: $icon-light-color;
    border-bottom-style: solid;
    border-bottom-color: $accent-color;
    border-bottom-width: 0.1rem;
    background-color: none !important;
  }

  .svg-btn.active svg {
    fill: $icon-light-color;
    stroke: #fff;
  }

  .td,
  .th {
    min-height: 5rem;
    padding: 0.5rem 1.1rem;
    align-items: center;
    border-right: 0.1rem solid $table-column-border-color;
    &:last-child {
      border-right: none;
    }
    &.svg-btn.expander:hover,
    &.svg-btn.expander:hover svg {
      fill: $accent-color;
    }
    &.svg-btn.reset-filter:hover,
    &.svg-btn.reset-filter:hover svg {
      fill: $accent-color;
    }
    &.expander,
    &.reset-filter {
      display: flex;
      align-items: center;

      &.svg-btn {
        display: block;
      }
    }
    &.clickable {
      cursor: pointer;
    }
    .expand-button:focus {
      background: none !important;
    }
    .sortable-column {
      display: flex;
      height: 100%;
      align-items: center;
      justify-content: inherit;
    }
    .filter-wrapper {
      div {
        div {
          display: flex;
          .form-group {
            width: 100%;
            .rbt {
              width: 100%;
              div {
                input {
                  width: 100% !important;
                }
              }
              .rbt-menu {
                width: 37rem !important;
              }
            }
          }
        }
      }

      &.active {
        color: $active-color;
      }
    }
  }
}
