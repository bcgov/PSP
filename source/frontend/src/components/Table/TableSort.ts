export type SortDirection = 'desc' | 'asc' | undefined;

export type TableSort<T extends object = object> = {
  // key (properties from the data object), value = sort direction
  [key in keyof T]?: SortDirection;
};
