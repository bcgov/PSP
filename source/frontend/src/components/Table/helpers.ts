import ITypeCode from 'interfaces/ITypeCode';
import { CellProps } from 'react-table';
import { formatMoney, formatNumber, prettyFormatDate } from 'utils';

/**
 * These helper methods below provide ways to render common values on a Table cell;
 *  - dates
 *  - money
 *  - percentages
 */

export const renderDate = <T extends object>({ cell: { value } }: CellProps<T, string>) =>
  prettyFormatDate(value);

export const renderMoney = <T extends object>({ cell: { value } }: CellProps<T, string>) =>
  formatMoney(value as any);

export const renderPercentage = <T extends object>({ cell: { value } }: CellProps<T, string>) =>
  `${formatNumber(value as any, 0, 2)}%`;

export function renderBooleanAsYesNo({ value }: CellProps<any, boolean>) {
  return value ? 'Y' : 'N';
}

export function renderTypeCode({ value }: CellProps<any, ITypeCode<any>>) {
  return value?.description ?? '';
}
