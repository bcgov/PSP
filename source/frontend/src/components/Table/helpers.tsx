import Api_TypeCode from 'models/api/TypeCode';
import { CellProps } from 'react-table';
import { formatMoney, formatNumber, prettyFormatDate, stringToFragment } from 'utils';

/**
 * These helper methods below provide ways to render common values on a Table cell;
 *  - dates
 *  - money
 *  - percentages
 */

export const renderDate = <T extends object>({
  cell: { value },
}: CellProps<T, string | Date | undefined>) => stringToFragment(prettyFormatDate(value));

export const renderMoney = <T extends object>({
  cell: { value },
}: CellProps<T, number | '' | undefined>) => stringToFragment(formatMoney(value));

export const renderPercentage = <T extends object>({ cell: { value } }: CellProps<T, number>) =>
  stringToFragment(`${formatNumber(value, 0, 2)}%`);

export const renderBooleanAsYesNo = ({ value }: CellProps<any, boolean | undefined>) =>
  stringToFragment(value ? 'Y' : 'N');

export const renderTypeCode = ({ value }: CellProps<any, Api_TypeCode<any> | undefined>) =>
  stringToFragment(value?.description ?? '');
