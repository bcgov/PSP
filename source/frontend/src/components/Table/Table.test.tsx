import { act, render, RenderOptions, userEvent } from 'utils/test-utils';

import { ColumnWithProps, Table, TableProps } from '.';
import { IIdentifiedObject } from './Table';

const setSort = jest.fn();

interface test extends IIdentifiedObject {
  name: string;
}

const testColumns: ColumnWithProps<test>[] = [
  {
    Header: 'Column number',
    accessor: 'name',
    align: 'left',
    width: 20,
    maxWidth: 40,
  },
  {
    Header: 'Column name',
    accessor: 'id',
    align: 'left',
    width: 20,
    maxWidth: 40,
  },
];

const testData: test[] = [
  { name: 'one', id: 1 },
  { name: 'two', id: 2 },
  { name: 'three', id: 3 },
  { name: 'four', id: 4 },
  { name: 'five', id: 5 },
  { name: 'six', id: 6 },
];

// render component under test
const setup = <T extends IIdentifiedObject>(
  renderOptions: RenderOptions & { props?: Partial<TableProps<T>> },
) => {
  const { props: tableProps, ...rest } = renderOptions;
  const utils = render(
    <Table<T>
      {...tableProps}
      name={tableProps?.name ?? 'default'}
      columns={tableProps?.columns ?? (testColumns as any)}
      data={(tableProps?.data ?? testData) as any}
    />,
    {
      ...rest,
    },
  );
  const tableRows = utils.container.querySelectorAll('.table .tbody .tr-wrapper');
  return {
    ...utils,
    tableRows,
  };
};

describe('Generic table component', () => {
  beforeEach(() => {
    setSort.mockClear();
  });

  it('matches snapshot', async () => {
    const { asFragment } = setup({});
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a message when no matching records found', async () => {
    const { tableRows, getByText } = setup({ props: { data: [] } });

    expect(tableRows.length).toBe(0);
    expect(getByText('No rows to display')).toBeVisible();
  });

  describe('table pagination', () => {
    it('by default the first page is the active page', async () => {
      const { getByLabelText } = setup({ props: { pageSize: 5 } });

      const page1Button = getByLabelText('Page 1 is your current page');
      expect(page1Button.parentElement).toHaveClass('active');
    });

    it('it displays a maximum number of rows equal to the pagesize prop if manualPagination is false', async () => {
      const { tableRows } = setup({ props: { pageSize: 5, manualPagination: false } });

      expect(tableRows).toHaveLength(5);
    });

    it('it displays all rows if manualPagination is true', async () => {
      const { tableRows } = setup({ props: { pageSize: 5, manualPagination: true } });

      expect(tableRows).toHaveLength(6);
    });

    it('allows the current page to be changed', async () => {
      const { getByLabelText, getByText } = setup({ props: { pageSize: 5 } });

      const page2Button = getByLabelText('Page 2');
      act(() => {
        userEvent.click(page2Button);
      });
      expect(getByText('six')).toBeVisible();
      expect(page2Button.parentElement).toHaveClass('active');
    });

    it('displays multiple pages when the total number of rows exceed the page size', async () => {
      const { getByLabelText } = setup({ props: { pageSize: 5 } });

      expect(getByLabelText('Page 2')).toBeVisible();
    });

    it('hides the paging control when hidePagination is set', async () => {
      const { tableRows, queryByLabelText } = setup({
        props: { pageSize: 5, hidePagination: true },
      });

      const page1Button = queryByLabelText('Page 1 is your current page');
      expect(page1Button).toBeNull();
      expect(tableRows).toHaveLength(6);
    });

    it('does not change the page when manual pagination is set', async () => {
      const onRequestData = jest.fn();
      const { getByLabelText, tableRows } = setup({
        props: { pageSize: 5, manualPagination: true, onRequestData },
      });

      const page2Button = getByLabelText('Page 2');
      act(() => {
        userEvent.click(page2Button);
      });
      expect(tableRows).toHaveLength(6); //pagination is handled externally, the table component will ignore the page size prop.
      expect(onRequestData).toHaveBeenCalled();
    });

    it('page size works as expected when not using manual pagination', async () => {
      const { getByLabelText, getByTitle, container } = setup({
        props: { pageSize: 10, manualPagination: false },
      });

      await act(() => {
        userEvent.click(getByTitle('menu-item-5'));
      });
      const tableRows = container.querySelectorAll('.table .tbody .tr-wrapper');
      expect(tableRows).toHaveLength(5);
      expect(getByLabelText('Page 2')).toBeVisible();
    });

    it('page size works as expected when changing page size and then changing page', async () => {
      const { getByLabelText, getByTitle, container } = setup({
        props: { manualPagination: false },
      });

      await act(() => userEvent.click(getByTitle('menu-item-5')));
      const tableRows = container.querySelectorAll('.table .tbody .tr-wrapper');
      expect(tableRows).toHaveLength(5);
      await act(() => userEvent.click(getByLabelText('Page 2')));
      expect(getByLabelText('Page 2 is your current page')).toBeVisible();
    });

    it('page size works as expected when using manual pagination', async () => {
      const onPageSizeChange = jest.fn();
      const { getByTitle } = setup({
        props: { pageSize: 10, manualPagination: true, onPageSizeChange },
      });

      act(() => {
        userEvent.click(getByTitle('menu-item-5'));
      });
      expect(onPageSizeChange).toHaveBeenCalledWith(5);
    });
  });

  it('can select only one row at a time when single select prop set', async () => {
    const setSelectedRows = jest.fn();
    const { getByTestId } = setup({
      props: { pageSize: 5, isSingleSelect: true, setSelectedRows, showSelectedRowCount: true },
    });

    const selectRowOneButton = getByTestId('selectrow-1');
    act(() => {
      userEvent.click(selectRowOneButton);
    });

    const selectRowTwoButton = getByTestId('selectrow-2');
    act(() => {
      userEvent.click(selectRowTwoButton);
    });

    expect(setSelectedRows).toHaveBeenNthCalledWith(1, [{ id: 1, name: 'one' }]);
    expect(setSelectedRows).toHaveBeenNthCalledWith(2, [{ id: 2, name: 'two' }]);
    expect(selectRowOneButton).toHaveAttribute('type', 'radio');
    expect(selectRowTwoButton).toHaveAttribute('type', 'radio');
  });

  it('can select multiple rows when isSingleSelect is false', async () => {
    const setSelectedRows = jest.fn();
    const { getByTestId } = setup({
      props: { pageSize: 5, isSingleSelect: false, setSelectedRows, showSelectedRowCount: true },
    });

    const selectRowOneButton = getByTestId('selectrow-1');
    act(() => {
      userEvent.click(selectRowOneButton);
    });
    const selectRowTwoButton = getByTestId('selectrow-2');
    act(() => {
      userEvent.click(selectRowTwoButton);
    });

    expect(setSelectedRows).toHaveBeenNthCalledWith(1, [{ id: 1, name: 'one' }]);
    expect(setSelectedRows).toHaveBeenNthCalledWith(2, [
      { id: 1, name: 'one' },
      { id: 2, name: 'two' },
    ]);
    expect(selectRowOneButton).toHaveAttribute('type', 'checkbox');
    expect(selectRowOneButton).toBeChecked();
    expect(selectRowTwoButton).toBeChecked();
  });
});
