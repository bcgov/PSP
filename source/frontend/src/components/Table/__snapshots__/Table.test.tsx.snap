// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic table component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  <div
    class="table"
    data-testid="default"
    role="table"
  >
    <div
      class="thead thead-light"
    >
      <div
        class="tr"
        role="row"
        style="display: flex; flex: 1 0 auto; min-width: 60px;"
      >
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Column number
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Column name
          </div>
        </div>
      </div>
    </div>
    <div
      class="tbody"
      role="rowgroup"
    >
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            one
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            1
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            two
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            2
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            three
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            3
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            four
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            4
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            five
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            5
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 60px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            six
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            6
          </div>
        </div>
      </div>
    </div>
  </div>
  .c0 {
  margin-top: 0.3rem;
}

<div
    class="row"
  >
    <div
      class="align-self-center col-auto"
    >
      <span>
        1 - 6 of  6
      </span>
    </div>
    <div
      class="ml-auto align-self-center col-auto"
    />
    <div
      class="align-self-center col-auto"
    >
      <div
        class="Menu-root"
      >
        <div
          class="Menu-button"
        >
          <div
            style="display: flex;"
          >
            <span
              class="c0"
            >
              Show
            </span>
            <input
              class="form-control form-control-sm"
              disabled=""
              style="width: 50px; margin-left: 10px; margin-right: 10px; background-color: white;"
              type="number"
              value="10"
            />
            <span
              class="c0"
            >
              Entries
            </span>
          </div>
        </div>
        <div
          class="Menu-items"
          style="width: 10.0rem;"
        >
          <div
            class="Menu-options scrollable list-group"
          >
            <div
              class="Menu-item list-group-item"
              title="menu-item-5"
            >
              5
            </div>
            <div
              class="Menu-item list-group-item"
              title="menu-item-10"
            >
              10
            </div>
            <div
              class="Menu-item list-group-item"
              title="menu-item-20"
            >
              20
            </div>
            <div
              class="Menu-item list-group-item"
              title="menu-item-50"
            >
              50
            </div>
            <div
              class="Menu-item list-group-item"
              title="menu-item-100"
            >
              100
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="align-self-center col-auto"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Previous page"
            class="page-link "
            rel="prev"
            role="button"
            tabindex="-1"
          >
            &lt;
          </a>
        </li>
        <li
          class="page-item active"
        >
          <a
            aria-current="page"
            aria-label="Page 1 is your current page"
            class="page-link"
            rel="canonical"
            role="button"
            tabindex="-1"
          >
            1
          </a>
        </li>
        <li
          class="page-item disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Next page"
            class="page-link "
            rel="next"
            role="button"
            tabindex="-1"
          >
            &gt;
          </a>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
